pkg/envoy/registry/services.go

```diff
@@ -27,34 +34,168 @@ func (e ExplicitProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service
 
 // KubeProxyServiceMapper maps an Envoy instance to services in a Kubernetes cluster.
 type KubeProxyServiceMapper struct {
-	KubeController k8s.Controller
+	kubeController k8s.Controller
+	kubeEvents     chan interface{}
+	servicesForCN  sync.Map // map[certificate.CommonName][]service.MeshService
 }
 
-// ListProxyServices maps an Envoy instance to a number of Kubernetes services.
-func (k *KubeProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service.MeshService, error) {
-	cn := p.GetCertificateCommonName()
+// NewKubeProxyServiceMapper initializes a KubeProxyServiceMapper with an empty cache.
+func NewKubeProxyServiceMapper(controller k8s.Controller) *KubeProxyServiceMapper {
+	return &KubeProxyServiceMapper{
+		kubeController: controller,
+		kubeEvents: events.GetPubSubInstance().Subscribe(
+			announcements.PodAdded,
+			announcements.PodUpdated,
+			announcements.PodDeleted,
+			announcements.ServiceAdded,
+			announcements.ServiceUpdated,
+			announcements.ServiceDeleted,
+		),
+	}
+}
 
-	pod, err := envoy.GetPodFromCertificate(cn, k.KubeController)
-	if err != nil {
-		return nil, err
+// Run starts updating the proxy-to-services cache based on k8s notifications.
+func (k *KubeProxyServiceMapper) Run(stop <-chan struct{}) {
+	// Populate the cache using the cluster's existing state. Otherwise if
+	// existing resources are not modified, no events will come through to
+	// update the cache.
+	for _, pod := range k.kubeController.ListPods() {
+		k.handlePodUpdate(pod)
```

Comment:
@eduser25, @shashankram I noticed the e2e test that does a controller restart was failing with this change since it wasn't accounting for pods/services that had already been created and were never updated after this started.

Doing one initial list of the pods to populate the cache fixes that particular test, but would we need this to happen at some regular interval? If so, what kind of interval do you think is reasonable?

Reply:
I didn't completely understand the problem. Is this change somehow having an adverse effect when the controller is restarted?

Reply:
Because the new osm-controller starts with a brand new (empty) cache, if the pods and services that are already in the mesh never get updated, no event ever comes through to let the mapper know that they exist and to cache them. Since nothing is cached, then the mapper incorrectly replies that the proxies for those pods back no services.

Reply:
Are you implying that restarting the controller pushes incorrect/incomplete config to the proxies? I don't think we should introduce interval based syncs for this. There should be a mechanism to force an initial sync when the controller starts, after which subsequent events should just work without any need to periodically sync the caches.

Reply:
I suppose it's possible (though unlikely) for a new pod to be created after the initial `ListPods()` here and before we start the loop where it listens for events. In that case though would those be events already be buffered in the channel once we make it to the loop since this is already subscribed? If so then it seems like this is ok as-is.